src/tracks/ui/BrushHandle.cpp

```diff
@@ -0,0 +1,567 @@
+         oldSnapState.outCoord == newSnapState.outCoord) )
+      return 0;
+
+   return RefreshCode::RefreshAll;
+}
+
+BrushHandle::BrushHandle
+      ( const std::shared_ptr<TrackView> &pTrackView, bool useSnap,
+        const TrackList &trackList,
+        const TrackPanelMouseState &st, const ViewInfo &viewInfo,
+        const std::shared_ptr<SpectralData> &mpSpectralData,
+        const int brushRadius)
+      : mpView{ pTrackView }
+      , mpSpectralData(mpSpectralData)
+{
+   const wxMouseState &state = st.state;
+   auto pTrack = pTrackView->FindTrack().get();
+   auto wt = dynamic_cast<WaveTrack *>(pTrack);
+   double rate = mpSpectralData->GetSR();
+
+   mRect = st.rect;
+   mBrushRadius = brushRadius;
+   mFreqUpperBound = wt->GetSpectrogramSettings().maxFreq - 1;
+   mFreqLowerBound = wt->GetSpectrogramSettings().minFreq + 1;
+   // Borrowed from TimeToLongSample
+   mSampleCountLowerBound = floor( pTrack->GetStartTime() * rate + 0.5);
+   mSampleCountUpperBound = floor( pTrack->GetEndTime() * rate + 0.5);
+}
+
+BrushHandle::~BrushHandle()
+{
+}
+
+namespace {
+   // Is the distance between A and B less than D?
+   template < class A, class B, class DIST > bool within(A a, B b, DIST d)
+   {
+      return (a > b - d) && (a < b + d);
+   }
+
+   inline double findMaxRatio(double center, double rate)
+   {
+      const double minFrequency = 1.0;
+      const double maxFrequency = (rate / 2.0);
+      const double frequency =
+            std::min(maxFrequency,
+                     std::max(minFrequency, center));
+      return
+            std::min(frequency / minFrequency, maxFrequency / frequency);
+   }
+}
+
+void BrushHandle::Enter(bool, AudacityProject *project)
+{
+
+}
+
+bool BrushHandle::Escape(AudacityProject *project)
+{
+   return false;
+}
+
+// Add or remove data accroding to the ctrl key press
+void BrushHandle::HandleTimeFreqData(long long ll_sc, wxInt64 freq) {
+   // Ignore the mouse dragging outside valid area
+   // We should also check for the available freq. range that is visible to user
+   if(ll_sc < mSampleCountLowerBound || ll_sc > mSampleCountUpperBound ||
+      freq < mFreqLowerBound || freq > mFreqUpperBound)
+      return;
+
+   if(mbCtrlDown)
+      mpSpectralData->removeTimeFreqData(ll_sc, freq);
+   else
+      mpSpectralData->addTimeFreqData(ll_sc, freq);
+}
+
+UIHandle::Result BrushHandle::Click
+      (const TrackPanelMouseEvent &evt, AudacityProject *pProject)
+{
+   using namespace RefreshCode;
+
+   const auto pView = mpView.lock();
+   if ( !pView )
+      return Cancelled;
+
+   wxMouseEvent &event = evt.event;
+   const auto sTrack = TrackList::Get( *pProject ).Lock( FindTrack() );
+   const auto pTrack = sTrack.get();
+   const WaveTrack *const wt =
+         static_cast<const WaveTrack*>(pTrack);
+   auto &trackPanel = TrackPanel::Get( *pProject );
+   auto &viewInfo = ViewInfo::Get( *pProject );
+
+   mMostRecentX = event.m_x;
+   mMostRecentY = event.m_y;
```

Comment:
I think most of these lines are deletable copies from SelectHandle?

But maybe the click should call HandleTimeFreqData once -- if you follow the suggestion elsewhere that the first pick (at least in the life of the SpectralData object, not necessarily the first pick in the life of each click-drag-release) should be at the exact time coordinate at the mouse-down point, estabilishing the time origin for rounding to hops.